etc/systemd/system-generators/zfs-mount-generator.in

```diff
@@ -168,6 +170,46 @@ process_line() {
       "${dataset}" >/dev/kmsg
   fi
 
+  # Minimal pre-requisites to mount a ZFS dataset
+  wants="zfs-import.target"
+  if [ -n "${p_encroot}" ] &&
+      [ "${p_encroot}" != "-" ] ; then
+    keyloadunit="zfs-load-key-$(systemd-escape "${p_encroot}").service"
+    if [ "${p_encroot}" = "${dataset}" ] ; then
+        pathdep=""
+      if [ "${p_keyloc%%://*}" = "file" ] ; then
+        pathdep="RequiresMountsFor=${p_keyloc#file://}"
+        keyloadcmd="@sbindir@/zfs load-key ${dataset}"
+      elif [ "${p_keyloc}" = "prompt" ] ; then
+  keyloadcmd="sh -c 'systemd-ask-password | @sbindir@/zfs load-key ${dataset}'"
+      else
+        printf 'zfs-mount-generator: (%s) invalid keylocation\n' \
+          "${dataset}" >/dev/kmsg
+      fi
+      cat > "${dest_norm}/${keyloadunit}" << EOF
+# Automatically generated by zfs-mount-generator
+
+[Unit]
+Description=Load ZFS key for ${dataset}
+SourcePath=${cachefile}
+Documentation=man:zfs-mount-generator(8)
+DefaultDependencies=no
+Wants=${wants}
+After=${wants}
+${pathdep}
+
+[Service]
+Type=oneshot
+RemainAfterExit=yes
+ExecStart=${keyloadcmd}
+ExecStop=@sbindir@/zfs unload=key ${dataset}
```

Comment:
This should be a dash instead of an equals sign, right?

Reply:
Dataset probably needs quoting in case it contains spaces? Plus probably in other places too.

Reply:
You're absolutely correct. I'll double check the rules for systemd units, they're a little different.

Reply:
I see that you are quoting ${dataset} with single quotes. So I wanted to know if a ZFS dataset can contain a single quote. It cannot.